onnx/src/ops/ml/tree.rs

```diff
@@ -0,0 +1,664 @@
+    ) -> Self {
+        Node::Branch(BranchNode { cmp, feature_id, value, true_id, false_id, nan_is_true })
+    }
+
+    pub fn new_leaf(start_id: usize, end_id: usize) -> Self {
+        Node::Leaf(LeafNode { start_id, end_id })
+    }
+}
+
+#[derive(Copy, Clone, Debug)]
+pub struct Leaf {
+    class_id: usize,
+    weight: f32,
+}
+
+impl Leaf {
+    pub fn new(class_id: usize, weight: f32) -> Leaf {
+        Leaf { class_id, weight }
+    }
+}
+
+pub trait AggregateFn: Default {
+    fn aggregate(&mut self, score: f32, total: &mut f32);
+
+    fn post_aggregate(&mut self, total: &mut f32) {}
+}
+
+#[derive(Clone, Copy, Default, Debug)]
+pub struct SumFn;
+
+impl AggregateFn for SumFn {
+    fn aggregate(&mut self, score: f32, total: &mut f32) {
+        *total += score;
+    }
+}
+
+#[derive(Clone, Copy, Default, Debug)]
+pub struct AvgFn {
+    count: usize,
+}
+
+impl AggregateFn for AvgFn {
+    fn aggregate(&mut self, score: f32, total: &mut f32) {
+        *total += score;
+        self.count += 1;
+    }
+
+    fn post_aggregate(&mut self, total: &mut f32) {
+        if self.count > 1 {
+            *total /= (self.count as f32);
+        }
+        self.count = 0;
+    }
+}
+
+#[derive(Clone, Copy, Default, Debug)]
+pub struct MaxFn;
+
+impl AggregateFn for MaxFn {
+    fn aggregate(&mut self, score: f32, total: &mut f32) {
+        *total = total.max(score);
+    }
+}
+
+#[derive(Clone, Copy, Default, Debug)]
+pub struct MinFn;
+
+impl AggregateFn for MinFn {
+    fn aggregate(&mut self, score: f32, total: &mut f32) {
+        *total = total.min(score);
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Aggregate {
+    Sum,
+    Avg,
+    Max,
+    Min,
+}
+
+impl Default for Aggregate {
+    fn default() -> Self {
+        Aggregate::Sum
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum PostTransform {
+    Softmax,
+    Logistic,
+    SoftmaxZero,
+    // Probit, // probit, especially multinomial, is p.i.t.a. - so let's ignore it for now
+}
+
+impl PostTransform {
+    pub fn apply(&self, scores: &mut ArrayViewMut1<f32>) {
+        match self {
+            PostTransform::Softmax => {
+                // TODO: stability w.r.t. nan/inf/-inf?
+                let max: f32 = scores
+                    .iter()
+                    .cloned()
+                    .max_by(|a, b| a.partial_cmp(&b).unwrap_or(Ordering::Equal))
+                    .unwrap_or(0.0);
+                let mut norm = 0.;
+                scores.map_inplace(|a| {
+                    *a = (*a - max).exp();
+                    norm += *a;
+                });
+                scores.map_inplace(|a| *a = *a / norm);
+            }
+            PostTransform::Logistic => {
+                scores.map_inplace(|a| {
+                    let v = 1. / (1. + f32::exp(-f32::abs(*a)));
+                    *a = if a.is_sign_negative() { 1. - v } else { v };
```

Comment:
I have a faster implementation somewhere in the guts of the Sigmoid operator (as a rational function, accurate to f32). I think I will move it to linalg (which I will rename tract-compute), so that we can use it anywhere it makes sense.

Reply:
It's probably less critical here than in neural networks, so no rush.

Reply:
Makes sense. I flipped around the codebase but didn't find it - I guess I've missed it somehow.

Maybe there should be a toolbox of common operations, some of them simple scalar functions, some of them dealing with vectors (low-level, without any tract glue, without any validation, without bounds-checking etc), implemented as fast as possible so they can be reused in different parts of the library - like softmax, logistic, etc.

Reply:
Yeah, that was the idea of tract-linalg. One of the key features is its ability to compile fast, in order to iterate quickly over tests and benches of implementation. But I got the name wrong. This goes beyond linear algebra.